pkg/raft/raft.go

```diff
@@ -1040,6 +1037,39 @@ func (r *raft) reset(term uint64) {
 	r.uncommittedSize = 0
 }
 
+func (r *raft) setTerm(term uint64) {
+	assertTrue(term >= r.Term, "term cannot regress")
+	r.Term = term
+}
+
+func (r *raft) setVote(vote pb.PeerID) {
+	assertTrue(r.Vote == None || r.Vote == vote, "cannot change vote")
+	r.Vote = vote
+}
+
+func (r *raft) resetVote() {
+	r.Vote = None
+}
```

Comment:
These two methods, if used incorrectly, can result in changing the vote: if we first `resetVote` (without bumping the term), and then `setVote` to a different value.

Should we make these methods stricter in this regard? We could reset all the things in `setTerm` if the term is bumped, and remove the other `reset` methods. Then we would have the guarantee that the `set` methods do ensure the vote/lead/etc are not changed.

The `resetLead` is probably a special case here. There is some code deliberately resetting the `lead` to nil, which introduces the vulnerability that it can be changed to a different lead. For this case, I think we should, instead of resetting the `lead = None`, remember a bit saying that we "revoked" this lead, and act on this bit accordingly.

Reply:
Something like:

```golang
func setTerm(term) {
	if term == r.Term {
		return
	}
	assert(term > r.Term)
	r.Vote = None
	r.Lead = None
	// etc
}

func setVote(id) {
	if id == r.Vote {
		return
	}
	assert(r.Vote == None)
	r.Vote = id
}

func resetLead() {
	assert(r.Lead != None)
	r.leadRevoked = true
}

...
```

Reply:
Good points. I've addressed this as you suggested.